### PR TITLE
Fix std::complex broken tests

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we fix some std::complex broken tests under `__GLIBCXX__`:
 - complex_divide_complex.pass.cpp
 - complex_times_complex.pass.cpp
 - acos.pass.cpp
 - exp.pass.cpp
 - tanh.pass.cpp

## complex_divide_complex.pass.cpp
- introduced additional define `_PSTL_TEST_COMPLEX_DIV_COMPLEX_BROKEN_GLIB_CXX` :
```C++
// According to https://gcc.gnu.org/develop.html#timeline use last known _GLIBCXX_ to check the version of libstdc++ : 20240426
#define _PSTL_TEST_COMPLEX_DIV_COMPLEX_BROKEN_GLIB_CXX (__GLIBCXX__ > 0 && __GLIBCXX__ <= 20240426)
```
- fixed broken test cases on host.
- due this fix they are also fixed in Kernel.
- reproduce this case:
	https://gcc.godbolt.org/z/56nsf5qWo - without fast math
	https://gcc.godbolt.org/z/hcKqsd59K - with fast math

## complex_times_complex.pass.cpp
- changed define `_PSTL_TEST_COMPLEX_TIMES_COMPLEX_BROKEN` :
```C++
#define _PSTL_TEST_COMPLEX_TIMES_COMPLEX_BROKEN (_PSTL_TEST_COMPLEX_OP_BROKEN || (__GLIBCXX__ > 0 && __GLIBCXX__ <= 20240426))
``` 
- fixed broken test cases on host.
- due this fix they are also fixed in Kernel.

## acos.pass.cpp
- introduced additional define `_PSTL_TEST_COMPLEX_ACOS_BROKEN_IN_KERNEL_GLIB_CXX` :
```C++
#define _PSTL_TEST_COMPLEX_ACOS_BROKEN_IN_KERNEL_GLIB_CXX (__SYCL_DEVICE_ONLY__ && __GLIBCXX__ > 0 && __GLIBCXX__ <= 20240426)
```
- fixed broken test cases in Kernel code.
- additional issue registered.

## exp.pass.cpp
```C++
#define _PSTL_TEST_COMPLEX_EXP_BROKEN_IN_KERNEL_GLIB_CXX (__SYCL_DEVICE_ONLY__ && __GLIBCXX__ > 0 && __GLIBCXX__ <= 20240426)
```
- fixed broken test cases in Kernel code.
- additional issue registered.

## tanh.pass.cpp
```C++
#define _PSTL_TEST_COMPLEX_TANH_BROKEN_IN_KERNEL_GLIB_CXX (__SYCL_DEVICE_ONLY__ && __GLIBCXX__ > 0 && __GLIBCXX__ <= 20240426)
```
- fixed broken test cases in Kernel code.
- additional issue registered.

## Other
Also in this PR we modified common code for tests of `std::complex` to have compile error (`static_assert`) if fast math ([1](https://simonbyrne.github.io/notes/fastmath/), [2](https://stackoverflow.com/questions/43140581/how-to-use-the-gcc-preprocessor-macro-fast-math)) is switched on.

### CODE REVIEW

File: test/xpu_api/numerics/complex.number/complex.transcendentals/asin.pass.cpp [hunk 1]
```diff
@@ -87,7 +87,7 @@ void test_edges()
 #if !_PSTL_TEST_COMPLEX_SIN_BROKEN
             assert(std::isnan(r.real()));
             assert(std::isinf(r.imag()));
-#ifndef _PSTL_ICC_TEST_COMPLEX_ASIN_MINUS_INF_NAN_BROKEN_SIGNBIT
+#ifndef _PSTL_ICC_TEST_COMPLEX_ASIN_MINUS_INF_NAN_BROKEN_SIGNBIT        // test case: 33
```
Comment (danhoeflinger) [inline test/xpu_api/numerics/complex.number/complex.transcendentals/asin.pass.cpp:90]: Since we never merged #1569, we still have all these `#ifndef` instances that need to be fixed to be `#if !`.  We could do this in a separate PR, or in this one. 
The tests affected are at least the list from this comment: https://github.com/oneapi-src/oneDPL/pull/1569#issuecomment-2100607272

File: test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp [hunk 2]
```diff
@@ -53,9 +53,11 @@ void test_edges()
                     assert(classify(r) == zero);
                     break;
                 case inf:
-#if !_PSTL_TEST_COMPLEX_DIV_COMPLEX_BROKEN && !_PSTL_TEST_COMPLEX_DIV_COMPLEX_BROKEN_GLIBCXX
+#if !_PSTL_TEST_COMPLEX_DIV_COMPLEX_BROKEN          // test case 92, 33
```
Comment (danhoeflinger) [inline test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp:56]: Are these test case comments meant to stay in the PR?  if so, perhaps we need more information about what they correspond to.

File: test/xpu_api/numerics/complex.number/complex.transcendentals/asin.pass.cpp (hunk 1 above)
Comment (SergeyKopienko) [reply test/xpu_api/numerics/complex.number/complex.transcendentals/asin.pass.cpp]: Fixed

File: test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp (hunk 2 above)
Comment (SergeyKopienko) [reply test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp:56]: I propose to save these comments, they are very useful and show the index or indexes of element from `testcases` array.

Comment (SergeyKopienko) [reply test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp]: Fixed

Comment (danhoeflinger) [reply test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp]: ok,  I'm trying to figure out how best to improve their utility to people without this "insider info".  Maybe instead of commenting index numbers to an array of test cases (which may be a maintenance challenge), we can just comment the actual test cases.  
So in this case (if I counted right):
`dpl::complex<double>( 0, 1.e+6) / dpl::complex<double>(+0., -1),`

Comment (SergeyKopienko) [reply test/xpu_api/numerics/complex.number/complex.ops/complex_divide_complex.pass.cpp]: Done

File: test/support/test_config.h [hunk 3]
```diff
@@ -130,34 +130,40 @@
...
+#define _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX (__GLIBCXX__ > 0 && __GLIBCXX__ <= 20210601)
+
+#define _PSTL_ICC_TEST_COMPLEX_ASIN_MINUS_INF_NAN_BROKEN_SIGNBIT          _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_COSH_MINUS_INF_MINUS_ZERO_BROKEN_SIGNBIT   _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_COSH_MINUS_ZERO_MINUS_ZERO_BROKEN_SIGNBIT  _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_POW_COMPLEX_COMPLEX_PASS_BROKEN_TEST_EDGES _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_POW_COMPLEX_SCALAR_PASS_BROKEN_TEST_EDGES  _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_POW_SCALAR_COMPLEX_PASS_BROKEN_TEST_EDGES  _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_NORM_MINUS_INF_NAN_BROKEN_TEST_EDGES       _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
+#define _PSTL_ICC_TEST_COMPLEX_POLAR_BROKEN_TEST_EDGES                    _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX
```
Comment (danhoeflinger) [inline test/support/test_config.h:143]: Just to confirm, these tests are broken on host AND in kernel where the ones below are broken only on the kernel?

File: test/support/test_config.h [hunk 4]
```diff
@@ -130,34 +130,40 @@
...
-#define _PSTL_ICC_TEST_COMPLEX_ASIN_MINUS_INF_NAN_BROKEN_SIGNBIT __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_COSH_MINUS_INF_MINUS_ZERO_BROKEN_SIGNBIT __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_COSH_MINUS_ZERO_MINUS_ZERO_BROKEN_SIGNBIT __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_POW_COMPLEX_COMPLEX_PASS_BROKEN_TEST_EDGES __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_POW_COMPLEX_SCALAR_PASS_BROKEN_TEST_EDGES __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_POW_SCALAR_COMPLEX_PASS_BROKEN_TEST_EDGES __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_NORM_MINUS_INF_NAN_BROKEN_TEST_EDGES __INTEL_LLVM_COMPILER
-#define _PSTL_ICC_TEST_COMPLEX_POLAR_BROKEN_TEST_EDGES __INTEL_LLVM_COMPILER
+// According to https://gcc.gnu.org/develop.html#timeline use last known _GLIBCXX_ to check the version of libstdc++ : 20240426
+#define _PSTL_TEST_COMPLEX_OP_BROKEN_GLIBCXX (__GLIBCXX__ > 0 && __GLIBCXX__ <= 20210601)
```
Comment (danhoeflinger) [inline test/support/test_config.h:134]: it seems that the tests which use this generally used to be turned off with `__INTEL_LLVM_COMPILER`.  
Have you confirmed then that this is an issue with stdlibc++?  

In other words, with libc++ do these work?  how about MSVC STL?

File: test/support/test_config.h (hunk 3 above)
Comment (SergeyKopienko) [reply test/support/test_config.h:143]: They are broken at least on host so I assume that inside Kernel they shouldn't work too.

File: test/support/test_config.h (hunk 4 above)
Comment (SergeyKopienko) [reply test/support/test_config.h:134]: I have checked under MS STL on my local machine ant it's working fine.

**Review by danhoeflinger — APPROVED** [review]
LGTM.

**Comment by SergeyKopienko** [comment]
@danhoeflinger, your re-approve required: there were merge conflicts after your PR with my changes.

**Review by danhoeflinger — APPROVED** [review]
